File: 0239-sliding-window-maximum/0239-sliding-window-maximum.py
from collections import deque

class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        queue = deque()
        result = []

        for i, n in enumerate(nums):
            while queue and nums[queue[-1]] < n:
                queue.pop()
            
            queue.append(i)
            
            if queue[0] == i - k:
                queue.popleft()

            if i >= k - 1:
                result.append(nums[queue[0]])
            
        return result

File: 0239-sliding-window-maximum/0239-sliding-window-maximum.rs
use std::collections::VecDeque;

impl Solution {
    pub fn max_sliding_window(nums: Vec<i32>, k: i32) -> Vec<i32> {
        let mut queue = VecDeque::new();
        let mut result: Vec<i32> = Vec::new();

        for (i, n) in nums.iter().enumerate() {
            while !queue.is_empty() && nums[*queue.back().unwrap()] < *n {
                queue.pop_back();
            }
            queue.push_back(i);
            if *queue.front().unwrap() + k as usize == i {
                queue.pop_front();
            }

            if i >= k as usize - 1{
                result.push(nums[*queue.front().unwrap()]);
            }
        }
        result
    }
}